Index: src/org/hibernate/mapping/SimpleValue.java
===================================================================
--- src/org/hibernate/mapping/SimpleValue.java	(revision 16560)
+++ src/org/hibernate/mapping/SimpleValue.java	(working copy)
@@ -188,8 +188,8 @@
 	}
 	
 	public boolean isIdentityColumn(Dialect dialect) {
-		return IdentifierGeneratorFactory.getIdentifierGeneratorClass(identifierGeneratorStrategy, dialect)
-				.equals(IdentityGenerator.class);
+        return IdentityGenerator.class.isAssignableFrom(
+            IdentifierGeneratorFactory.getIdentifierGeneratorClass(identifierGeneratorStrategy, dialect)); 
 	}
 
 	/**
